Create a Dockerfile for running the Python functional tests

davehunt opened the pull request:
To build the Docker image:

```
cp docker/dockerfiles/bedrock_functional_tests Dockerfile
docker build -t bedrock_functional_tests .
rm Dockerfile
```

To run the tests in Docker containers using Selenium Grid:

```
docker run -d --name selenium-hub selenium/hub:2.48.2
docker run -d --link selenium-hub:hub selenium/node-firefox:2.48.2
docker run -v `pwd`/results:/app/results \
--link selenium-hub:hub \
-e SELENIUM_HOST=hub \
bedrock_functional_tests
```

For additional nodes run several `node-firefox` containers by repeating the command to run it.

To run the tests using Sauce Labs:

```
docker run -v `pwd`/results:/app/results \
-e DRIVER=SauceLabs \
-e SAUCELABS_USERNAME=[user] \
-e SAUCELABS_API_KEY=[key] \
-e BROWSER_NAME=Firefox \
-e PLATFORM="Windows 10" \
-e SCREEN_RESOLUTION=1280x1024 \
bedrock_functional_tests
```

Replace `[user]` and `[key]` with the values provided by Sauce Labs.

You will find the HTML report and JUnit results in the `results` folder on your local machine.

@glogiotatidis could you take a look at this and provide feedback? I'd likely need some help hooking this up in Jenkins.

glogiotatidis commented on requirements/test_redirects.txt:
```diff
@@ -0,0 +1,9 @@
+-r dev.txt
+-r test.txt
+
+# sha256: -WfKOb25jhYpmmnEWpRMXUNFOTYV7WRwux5iyjUGv0E
```
Both `braceexpand` and `pytest-django` are in `test.txt`

glogiotatidis commented on requirements/test_redirects.txt:
```diff
@@ -0,0 +1,9 @@
+-r dev.txt
```
Do we need many packages from `dev.txt`? Does it make sense to add them in `test.txt` and don't include `dev.txt` here?

davehunt replied:
The redirect tests do, as they use `pytest-django` that at least requires `Django` itself. I didn't really want to add `dev.txt` as a requirement for the functional tests as they only currently run against an independently running instance of Bedrock.

glogiotatidis replied:
OK. I realized that this is not installed in docker and only used in travis, correct?

davehunt replied:
That's correct, although I suspect at some point we may want to include the redirect tests.

pmclanahan replied:
Yes! :+1:

glogiotatidis commented on docker/dockerfiles/bedrock_functional_tests:
```diff
@@ -0,0 +1,28 @@
+COPY . /app
+
+# Run the tests
+CMD ["/app/bin/run-functional-tests.sh"]
```
This is fine here. For an extra optimization step it can be moved after `WORKDIR`. It only defines the command to run on `docker run` and it doesn't have to be on the end of the file.

davehunt replied:
Ah, I was thinking of it more like a script. :)

glogiotatidis commented on bin/run-functional-tests.sh:
```diff
@@ -0,0 +1,41 @@
+fi
+
+CMD="${CMD} ${TESTS_PATH}"
+echo "Running: ${CMD}" && eval ${CMD}
```
I like to use `set -xe` at the start of my bash scripts. `-x` prints the commands before executing them, similarly to what you do here with `Running: ${CMD}` and `-e` exits immediatelly after a failed cmd.

davehunt replied:
+1 thanks

glogiotatidis commented on docker/dockerfiles/bedrock_functional_tests:
```diff
@@ -0,0 +1,28 @@
+
+WORKDIR /app
+
+# Defaults
```
What do you think about deduplicating the defaults and keeping in the dockerfile only the docker related defaults?

davehunt replied:
Yeah, I considered that myself. :)

glogiotatidis commented on docker/dockerfiles/bedrock_functional_tests:
```diff
@@ -0,0 +1,28 @@
+FROM python:2
```
nit: `python:2-slim` is a lighter img

davehunt replied:
nice

glogiotatidis commented:
Amazing work @davehunt ! I'm super happy that selenium provides it's own docker images and we can easily build upon that. 

We could possibly automate the start and stop of selenium + node + tests using docker-compose but it's probably good enough as is.

Re CI:

This is the new functional tests suite. Should we run this after unit tests step? Do we need locales installed for these tests to pass?

davehunt commented:
Thanks @glogiotatidis. I've addressed your review comments. Let me know if it looks good now and if so I'll go ahead and squash in preparation for merge.

> This is the new functional tests suite. Should we run this after unit tests step? Do we need locales installed for these tests to pass?

I think we should run this after the unit tests. Most of the tests expect the en-US locale (and explicitly target it) however there is at least one test that switches the current language on the home page. We could either include this but require l10n, or exclude this test by indicating a 'sanity' suite for the build pipeline. We could also mark any tests requiring l10n and exclude those easily enough.

/cc @alexgibson @jgmize @pmclanahan

alexgibson commented:
I'm going to open a PR shortly that will mark specific tests as sanity tests (we can discuss what to mark and what not to mark in that PR). I can make sure we leave out any tests that touch l10n or specific locales if that helps here.

glogiotatidis commented:
> I'm going to open a PR shortly that will mark specific tests as sanity tests (we can discuss what to mark and what not to mark in that PR). I can make sure we leave out any tests that touch l10n or specific locales if that helps here.

Yes that would be helpful thank you. 

The reason is that we include l10n as the very last building step of bedrock.

glogiotatidis commented:
Nice work @davehunt :cake: r+

alexgibson commented:
Opened #3476 to add markers for sanity tests

glogiotatidis commented:
I'm available to create the new pipeline step when this and #3476 get merged.

davehunt commented:
I've added support for a `MARK_EXPRESSION` environment variable. In the build pipeline we can set this to `sanity` to restrict test collection to just the sanity tests.

davehunt commented:
Squashed commits ready for merge.

glogiotatidis commented:
Thanks @davehunt :dizzy:
